health-events-analyzer/pkg/parser/parser.go

```
@@ -0,0 +1,204 @@
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"strconv"
+	"strings"
+
+	datamodels "github.com/nvidia/nvsentinel/data-models/pkg/model"
+	protos "github.com/nvidia/nvsentinel/data-models/pkg/protos"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+type Parser struct {
+	Event datamodels.HealthEventWithStatus
+}
+
+// parseSequenceString converts a criteria map into a BSON document for MongoDB queries
+func (p *Parser) ParseSequenceString(criteria map[string]any) (bson.D, error) {
+	doc := bson.D{}
+
+	for key, value := range criteria {
+		// Non-string values are appended as-is
+		strVal, isString := value.(string)
+
+		if !isString {
+			doc = append(doc, bson.E{Key: key, Value: value})
+			continue
+		}
+
+		// "this." reference → resolve from current event
+		if strings.HasPrefix(strVal, "this.") {
+			fieldPath := strings.TrimPrefix(strVal, "this.")
+			resolvedValue, err := p.getValueFromPath(fieldPath)
+
+			if err != nil {
+				return doc, fmt.Errorf("error in getting value from path: %w", err)
+			}
+
+			doc = append(doc, bson.E{Key: key, Value: resolvedValue})
+
+			continue
+		}
+
+		// JSON object string representing MongoDB operator (e.g. '{"$ne":"x"}')
+		if strings.HasPrefix(strVal, "{") && strings.HasSuffix(strVal, "}") {
+			var operatorMap map[string]any
+			if err := json.Unmarshal([]byte(strVal), &operatorMap); err != nil {
+				slog.Warn("Failed to parse MongoDB operator string, treating as literal", "string", strVal, "error", err)
```

Won't this cause the query to fail? Wouldn't it be better to fail fast here?